x/ibc/04-channel/keeper/timeout.go

```diff
@@ -62,56 +64,81 @@ func (k Keeper) TimeoutPacket(
 		)
 	}
 
-	if packet.GetDestPort() != channel.Counterparty.PortID {
-		return nil, sdkerrors.Wrapf(
-			types.ErrInvalidPacket,
-			"packet destination port doesn't match the counterparty's port (%s ≠ %s)", packet.GetDestPort(), channel.Counterparty.PortID,
-		)
-	}
-
+	// check that timeout height has passed on the other end
 	if proofHeight < packet.GetTimeoutHeight() {
 		return nil, types.ErrPacketTimeout
 	}
 
+	// check that packet has not been received
 	if nextSequenceRecv >= packet.GetSequence() {
-		return nil, sdkerrors.Wrap(
-			types.ErrInvalidPacket,
-			"packet already received",
-		)
+		return nil, sdkerrors.Wrap(types.ErrInvalidPacket, "packet already received")
 	}
 
 	commitment := k.GetPacketCommitment(ctx, packet.GetSourcePort(), packet.GetSourceChannel(), packet.GetSequence())
+
+	// verify we sent the packet and haven't cleared it out yet
 	if !bytes.Equal(commitment, types.CommitPacket(packet.GetData())) {
-		return nil, sdkerrors.Wrap(
-			types.ErrInvalidPacket,
-			"packet hasn't been sent",
-		)
+		return nil, sdkerrors.Wrap(types.ErrInvalidPacket, "packet hasn't been sent")
 	}
 
-	var ok bool
+	consensusState, found := k.clientKeeper.GetConsensusState(
+		ctx, connectionEnd.GetClientID(), proofHeight,
+	)
+	if !found {
+		return nil, clienttypes.ErrConsensusStateNotFound
+	}
+
+	var err error
 	switch channel.Ordering {
-	case types.ORDERED:
-		ok = k.connectionKeeper.VerifyMembership(
+	case exported.ORDERED:
+		// check that the recv sequence is as claimed
+		err = k.connectionKeeper.VerifyNextSequenceRecv(
 			ctx, connectionEnd, proofHeight, proof,
-			types.NextSequenceRecvPath(packet.GetDestPort(), packet.GetDestChannel()),
-			sdk.Uint64ToBigEndian(nextSequenceRecv),
+			packet.GetDestPort(), packet.GetDestChannel(), nextSequenceRecv,
+			consensusState,
 		)
-	case types.UNORDERED:
-		ok = k.connectionKeeper.VerifyNonMembership(
+	case exported.UNORDERED:
+		err = k.connectionKeeper.VerifyPacketAcknowledgementAbsence(
 			ctx, connectionEnd, proofHeight, proof,
-			types.PacketAcknowledgementPath(packet.GetSourcePort(), packet.GetSourceChannel(), packet.GetSequence()),
+			packet.GetDestPort(), packet.GetDestChannel(), packet.GetSequence(),
+			consensusState,
 		)
 	default:
 		panic(sdkerrors.Wrapf(types.ErrInvalidChannelOrdering, channel.Ordering.String()))
 	}
 
-	if !ok {
-		return nil, sdkerrors.Wrap(types.ErrInvalidPacket, "packet verification failed")
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "packet verification failed")
 	}
 
+	// NOTE: the remaining code is located on the TimeoutExecuted function
 	return packet, nil
 }
 
+// TimeoutExecuted deletes the commitment send from this chain after it verifies timeout
+func (k Keeper) TimeoutExecuted(ctx sdk.Context, packet exported.PacketI) error {
```

Comment:
I'm still a little confused as to the intended usage pattern here, cc @mossid - where is this / would this be called?

Reply:
Interested in this as well.